matrix_sdk_base/src/client.rs

```diff
@@ -493,6 +504,43 @@ impl BaseClient {
                         _ => (),
                     }
 
+                    if let Some(context) = &mut push_context {
+                        self.update_push_room_context(context, user_id, room_info, changes)
+                            .await;
+                    } else {
+                        push_context = self.get_push_room_context(room, room_info, changes).await?;
+                    }
+
+                    if let Some(context) = &push_context {
+                        let actions: Vec<Action> = push_rules
+                            .get_actions(&raw_event, &context)
+                            .cloned()
+                            .collect();
+
+                        if actions.iter().any(|a| matches!(a, Action::Notify)) {
+                            let notification = Notification::new(
+                                actions,
+                                raw_event,
+                                false,
+                                room_id.clone(),
+                                SystemTime::now(),
+                            );
+
+                            match changes.notifications.get_mut(room_id) {
```

Comment:
Let's move this into the `StateChanges` struct, something like `add_notifications()`.